Fix genesis root hash

David-Petrov opened the pull request:
In summary:
- fixes state root hash calculations;
- finishes up what @ablax had started with the `GenesisBlockHash`;
- some moving and repackaging;
- splitting of `InsertTrieBuilder` into a `TrieStructureFactory` and `InsertTrieBuilder` (the proper database related part of it);
- in-depth refactor and extension of `ChainSpec` (in-memory interpretation of the spec + JSON (de/)serializability + `implements Serializable` completely);
- fix for an unrelated issue regarding Spring beans (refer [13f935e](https://github.com/LimeChain/Fruzhin/pull/292/commits/13f935e557bc56721735b805f671b44c66056709))

Closes #289

Commit:
refactor!: move old trie into a separate subpackage

Commit:
refactor!: move `DecodedNode` into a separate package

out of `trie.structure` since it's not directly related

Commit:
fix!: node encoding corner case

was causing wrong root hash calculations

Commit:
fix!:

- cleaned up genesis block hash calculations (after fixing the state root hash calculations)

- extracted the trie building logic from `InsertTrieBuilder` into `TrieStructureFactory`. Now, `InsertTrieBuilder` only constructs the DTOs necessary to persist the trie into the database;

- for the purpose of building the trie, refactored `ChainSpec`, renaming it into `RawChainSpec` (a deserialized json file) and introducing the new `ChainSpec` class which is a parsed chain spec for easier in-memory work with the data... All the places that were previously using the raw chain spec are still using it raw, since that requires a bigger refactor, but had to go over and change a line in ~10 places. This also necessitated some refactoring of the beans involving `ChainSpec`;

- rewrote the tests around `ChainSpec`;

- small fixes (some annotations; some methods made static) around the TrieStructure;

- added integration tests for the state root hash calculation for each supported chain.

Commit:
feat:

- merge `dev`;

- also, introduce the forgotten uncommited parts of `RuntimeVersion` enum.

stoyan-lime commented on src/main/java/com/limechain/chain/spec/ChainSpec.java:
```diff
@@ -0,0 +1,25 @@
+
+import lombok.Getter;
+
+// Serves as a wrapper around the raw JSON parsed chain spec
```
Use javadoc instead of comments.

stoyan-lime commented on src/main/java/com/limechain/chain/spec/Genesis.java:
```diff
@@ -0,0 +1,45 @@
+public class Genesis {
+    private final Map<ByteString, ByteString> top;
+    // TODO: Add child storage support
+    // private Map<ByteString, ...?> childrenDefault;
```
Remove commented field.

Commit:
fix(build): remove exception message assertion

Commit:
chore(sonarlint): remove code smells

Commit:
chore(sonarlint): mark as utility class

temporarily, make `TrieStructureFactory` an utility class to avoid code smells and be consistent with Java's intentionality requirements... otherwise, I'm not sure if this is the best outlook for this class, it will most probably be refactored.

stoyan-lime commented on src/main/java/com/limechain/client/FullNode.java:
```diff
@@ -39,17 +43,20 @@ public class FullNode implements HostNode {
+
+            TrieStructure<NodeData> trie = TrieStructureFactory.buildFromKVPs(genesisStorage.getTop(), stateVersion);
+            List<InsertTrieNode> dbSerializedTrieNodes = new InsertTrieBuilder(trie).build();
+            insertStorage(db, dbSerializedTrieNodes, InsertTrieBuilder.STATE_VERSION); //Todo: calculate state version
```
The state version passed here is different than the one set above.

David-Petrov replied:
True, fixed this.

Commit:
refactor: remove unused hardcoded STATE_VERSION int

stoyan-lime commented on src/main/java/com/limechain/chain/spec/Genesis.java:
```diff
@@ -0,0 +1,46 @@
+    Genesis(RawChainSpec rawChainSpec) {
+        Map<ByteString, ByteString> parsedTopStorage = new HashMap<>();
+
+        Function<String, ByteString> parser =
```
You can extract this in a private method.

Commit:
chore(sonarlint): remove unnecessary local variables

Commit:
refactor(chainspec)!: polish previous quick and dirty chainspec solution

- merge the introduced two classes `ChainSpec` and `ChainSpecRaw` into a single `ChainSpec`, combining deserialization / serialization logic and in-memory representation using some Jackson customization

Commit:
refactor(chainspec)!: introduce `ChainType` enum

Commit:
refactor(chainspec)!: introduce `TelemetryEndpoint` record

Commit:
FIX!: disambiguate sync machine beans

completely unrelated to this PRs work, but alleviates an issue which was preventing the RpcApp to start

Commit:
chore: remove arguably obsolete TODO

Commit:
feat(chainspec): introduce `PropertyValue` type

- see the javadoc of the class

- in summary: values can be arbitrary and we needn't interpret them for the host, we'll only forward them as-is (again, in JSON) via RPC calls

georg-getz commented on src/main/java/com/limechain/chain/ChainService.java:
```diff
@@ -50,29 +47,29 @@ protected void initialize(HostConfig hostConfig) {
-        } catch (IOException e) {
-            throw new ChainServiceInitializationException(e);
+                repository.save(DBConstants.GENESIS_KEY, this.chainSpec);
+                log.log(Level.FINE, "Saved chain spec to database");
```
I'm thinking that we almost always want to be avoiding anything below `INFO` level. The main idea of fine/finer/finest is to be used for logs that would spam the output and we only very occasionally would like to see being printed in order to debug, whereas messages like this one I think should be at least `INFO` level

David-Petrov replied:
No probs, I've not changed that. 😄 Will now.

Commit:
chore: add javadocs

georg-getz commented on src/test/java/com/limechain/chain/StateRootHashesIntegrationTest.java:
```diff
@@ -0,0 +1,40 @@
+
+class StateRootHashesIntegrationTest {
+
+    // TODO: Refactor these hardcoded paths after we refactor to extract a logical ChainSpec module
```
It is actually not uncommon for tests to have hardcoded setups like this one. Given that it doesn't take that much space this TODO could be omitted.

David-Petrov replied:
I was bothered by the fact that tests share the same resources as the production environment... Ideally, I wanted to abstract the actual paths and bind them to the enums, but the problem is that they're configurable via `application.properties`, which isn't loaded at the time the enum instances are initialized. 

Anyways, to avoid the shared resources, perhaps cloning only the genesis data into separate files under `test/resources` and using that independent dump could be an option? Genesis data never changes, after all... and decoupling the test resources from the production resources seems like a fine idea. 

Anyway, I've removed the todo. If we decide it's worth it, we might go that extra bit.

georg-getz replied:
The concern you're raising has a way bigger scope than just this test so the TODO would be better positioned more outwardly. What I mean is that maybe it's not a bad idea for us to have a TODO.md where have improvement suggestions such as this one that's maybe in root folder so it's more easily viewable by all. 

Anyway this would require hearing the other team members opinions so it's a discussion item for tomorrow given that we are past working hours.

Commit:
refactor!: hide `childrenDefault` from `Genesis`

Commit:
chore: change log level

Commit:
chore: remove insignificant TODO

Commit:
chore: convert unused code to a NOTE
